### DISCUSSION TITLE
Why PJFNK calls computeQpResidual() during each linear step?

### DISCUSSION THREAD

**Discussion by Richard-happy** [idaholab/moose#23881] [Q&A General] [answered]
From Krylov methods, what we need is vectors like $\{-R_h, J(-R_h), J^2(-R_h),......\}$. So it's OK just call computeQpResidual() once each nonlinear. 

From [https://mooseframework.inl.gov/workshop/index.html#/9/10](url) ,it says **During each linear step of PJFNK, the computeQpResidual method is called to approximate the action of the Jacobian on the Krylov vector.** I'm not understand this.
Dose computeQpResidual() not mean computing $-R_h$?

**Comment by GiudGiud** [comment, marked as answer]
Hello

if you compute the residual twice, once with a slightly offset solution vector, you can approximate the action of the Jacobian 

Guillaume

**Reply by Richard-happy** [reply]
understand .thank you ,GiudGiud
